### PR TITLE
Writing models

### PR DESCRIPTION
**Pull request by DpEpsilon**
The thing I did at the hackathon with automap was on hard drugs.

The SQLAlchemy way looks much better, though I consider the table for storing old usernames a bit ugly. When you want to commit a Student, you need to ```db.session.add(student.username_object)``` as well as ```db.session.add(student)``` before running ```db.session.commit()```.

There should be a nice way to integrate these changes with whatever Grundo has done with logins.

### CODE REVIEW

File: aioregister/models.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+from flask import Flask
+from flask.ext.sqlalchemy import SQLAlchemy
+
+db = SQLAlchemy()
+
+class School(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(15), nullable=False, unique=True)
```
Comment (bgbnbigben) [inline aioregister/models.py]: uhh. surely if we're keeping a separate username object this should also be on that table?

Comment (bgbnbigben) [inline aioregister/models.py]: Wait, no, what's an "old" username?

**Comment by openendings** [comment]
Ben: reassigning to you because you seem to have some idea wtf is going on :P

**Comment by bgbnbigben** [comment]
@openendings So you could just add a `session.add(self.username_object)` to `Student`'s `__init__` method, or alternatively set up a listener like so:

    Session = scoped_session(sessionmaker())

    @event.listens_for(mapper, 'init')
    def auto_add(target, args, kwargs):
        Session.add(target)

**Comment by openendings** [comment]
@DpEpsilon README.md is out of date; please fix.

**Comment by DpEpsilon** [comment]
@bgbnbigben Robert said that if students were deleted, we don't want to reuse usernames, and so he suggested a separate table which would retain usernames of deleted students.

**Comment by bgbnbigben** [comment]
So I wasn't there and thus am hazy on details, but if a student registers for the AIO and is subsequently deleted, why do we care what their username was? Unless we're allowing them to keep their usernames for multiple years, which likely leads to more possible headaches with changing schools or etc.

**Comment by openendings** [comment]
TL;DR: we assume the data here is being constantly (idempotently) ported
into the AIO CMS instance. To prevent the import scripts from having to
deal with deletions (and all the accompanying timestamp related BS), we
treat usernames as permanently canonical.

**Comment by bgbnbigben** [comment]
There's still a puzzle piece I'm missing, but whatever, if you guys are happy with this design then I'll assume it was discussed and made sense at the time, so I'll defer to you.

I'll commit a listener that does The Right Thing on student creation.

**Comment by bgbnbigben** [comment]
Wait no I undefer; why is the best option not have a "deleted" table, and then just whenever a student deletes their account the uname goes into that table? CMS then gets a list of deleted + a list of everything in aio-register, deletes the "deleted" and does the straightforward set-subtraction to tell what it needs to add?

This way your consistent + idempotent push to AIO-CMS is exactly that (since you clear the deleted table on every push), you don't care as much about forever-uniqueness of names and everything Just Works, right?

**Comment by openendings** [comment]
"clear the deleted table on every push" -- i guess we were concerned about
incurred risk of clearing such a table without actually having the
operation succeed on the CMS end.

**Comment by bgbnbigben** [comment]
Sure, if you're worried about scorched-earth you could always just have CMS ack every deleted element or w/e. It just feels like we're using an ugly hack to avoid having to actually deal with data. *shrugs*, whatever, I'll just commit the listener.

**Comment by openendings** [comment]
Nah, your idea is probably better. This was kind of "first working solution
we came up with that seemed to address the issues", but if
record-of-deletion table is a more sensible idiom, we should go with that

**Comment by bgbnbigben** [comment]
I mean, the current way works, it just seems insane to have to keep track of what are explicitly throwaway usernames.

**Comment by openendings** [comment]
I agree, though I should note that our solution (main table foreign!primary
keys out to a larger one) is not significantly more heavyweight than (main
table + separate "dead user names" table with an implicit guarantee of no
overlap).

Having a table containing all deleted *and* active user names means we get
a tiny bit of 'type safety' from foreign key-ing to it. It's still quite
easy to detect which usernames are "deleted" by set subtraction.

**Comment by bgbnbigben** [comment]
ofc; they both achieve the same task and are isomorphic, one is just more mental bandwidth IMO, because it's not immediately obvious why that table exists. I don't mean to suggest the current system doesn't work -- it clearly does!

**Comment by bgbnbigben** [comment]
if I could have a sanity check on this that would be great

**Comment by bgbnbigben** [comment]
......github only allows one assignee per CL? >:C

**Comment by openendings** [comment]
"on CMS's 200 response we then nuke that table"
> "on CMS's 200 response we nuke everything we sent in that request" FTFY :P 

Yeah, looks good :)

**Comment by bgbnbigben** [comment]
Oh, yeah, in my head it was a magical atomic operation. :P 

I don't see a reason to limit password to ascii; I'd rather go wide than narrow. Plus, this way users can set their passwords to

a̰͛l͚̓l͎̓ ̥͊h̯͘â̼ì̭ļ̛ ̟͝c͇̀t̡̑h̩͛ụ́l͇̽h̠̒ų̔

and we can't stop them ^_^

**Comment by bgbnbigben** [comment]
(more specifically, if your keyboard layout is non-qwerty/some international layout, you might have non-ascii buttons that you want to use and unicode is an easier encoding to deal with than <insert country here>)

**Comment by openendings** [comment]
More options for summoning Ancient Ones is sufficient reason for anything
^_^

(...well, anything besides PHP, let's not go overboard here)
